### PR TITLE
[core] Retry if snapshot commit hint failed.

### PR DESCRIPTION
**Pull request by leaves12138**
### Purpose

Maybe over-writer file will fail, so it's better to add some retries.

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/fs/FileIOUtils.java [hunk 1]
```diff
@@ -44,4 +44,19 @@ public static FileIOLoader checkAccess(FileIOLoader fileIO, Path path, CatalogCo
         io.exists(path);
         return fileIO;
     }
+
+    public static void overwriteFileUtf8WithRetry(
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/fs/FileIOUtils.java:48]: inline this in commitHint

File: paimon-common/src/main/java/org/apache/paimon/fs/FileIOUtils.java [hunk 2]
```diff
@@ -44,4 +44,19 @@ public static FileIOLoader checkAccess(FileIOLoader fileIO, Path path, CatalogCo
...
+            FileIO fileIO, Path path, String content, int times) throws IOException {
+        int loopTime = times;
+        while (loopTime-- > 0) {
+            try {
+                fileIO.overwriteFileUtf8(path, content);
+                return;
+            } catch (IOException e) {
+                if (loopTime == 0) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/fs/FileIOUtils.java:56]: Sleep a random value here?

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java [hunk 3]
```diff
@@ -883,6 +884,6 @@ public void commitEarliestHint(long snapshotId) throws IOException {
 
     private void commitHint(long snapshotId, String fileName, Path dir) throws IOException {
         Path hintFile = new Path(dir, fileName);
-        fileIO.overwriteFileUtf8(hintFile, String.valueOf(snapshotId));
+        FileIOUtils.overwriteFileUtf8WithRetry(fileIO, hintFile, String.valueOf(snapshotId), 3);
```
Comment (xuzifu666) [inline paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java:887]: Hard code here? Or make it configurable better？

Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java]: ok

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java [hunk 4]
```diff
@@ -883,6 +885,22 @@ public void commitEarliestHint(long snapshotId) throws IOException {
...
         Path hintFile = new Path(dir, fileName);
-        fileIO.overwriteFileUtf8(hintFile, String.valueOf(snapshotId));
+        int loopTime = 3;
+        while (loopTime-- > 0) {
+            try {
+                fileIO.overwriteFileUtf8(hintFile, String.valueOf(snapshotId));
+                return;
+            } catch (IOException e) {
+                try {
+                    Thread.sleep(RANDOM.nextInt(1000) + 500);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java:895]: `ThreadLocalRandom.current()`

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java [hunk 5]
```diff
@@ -883,6 +885,22 @@ public void commitEarliestHint(long snapshotId) throws IOException {
...
+                } catch (InterruptedException ex) {
+                    // throw root cause
+                    throw new RuntimeException(e);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java:898]: Thread.currentThread().interrupt();

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**